sudoku/sudoku.py

```diff
@@ -0,0 +1,182 @@
+            return False
+
+    # Verifying subsquares
+    subsquare_coords = [(i, j) for i in range(m) for j in range(m)]
+    for r_scalar in range(m):
+        for c_scalar in range(m):
+            subsquare = [matrix[i + r_scalar * m][j + c_scalar * m] for i, j
+                         in subsquare_coords]
+            if set(subsquare) != solution:
+                print("Error in sub-square: ", subsquare)
+                return False
+
+    return True
+
+
+def main():
+    # Note: for the purposes of a code example, main() is written as a script 
+
+    # Read user input
+    if len(sys.argv) > 1:
+        filename = sys.argv[1]
+    else:
+        filename = "problem.txt"
+        print("Warning: using default problem file, '{}'. Usage: python "
+              "sudoku.py <sudoku filepath>".format(filename))
+
+    # Read sudoku problem
+    matrix = get_matrix(filename)
+
+    # Set up
+    n = len(matrix)          # Number of rows/columns in sudoku
+    m = int(math.sqrt(n))    # Number of rows/columns in sudoku subsquare
+    digits = range(1, n+1)
+
+    bqm = dimod.BinaryQuadraticModel({}, {}, 0.0, dimod.SPIN)
+
+    # Constraint: Each node can only select one digit
+    for row in range(n):
+        for col in range(n):
+            node_digits = [get_label(row, col, digit) for digit in digits]
+            one_digit_bqm = combinations(node_digits, 1)
+            bqm.update(one_digit_bqm)
+
+    # Constraint: Each row of nodes cannot have duplicate digits
+    for row in range(n):
+        for digit in digits:
+            row_nodes = [get_label(row, col, digit) for col in range(n)]
+            row_bqm = combinations(row_nodes, 1)
+            bqm.update(row_bqm)
+
+    # Constraint: Each column of nodes cannot have duplicate digits
+    for col in range(n):
+        for digit in digits:
+            col_nodes = [get_label(row, col, digit) for row in range(n)]
+            col_bqm = combinations(col_nodes, 1)
+            bqm.update(col_bqm)
+
+    # Constraint: Each sub-square cannot have duplicates
+    # Build indices of a basic subsquare
+    subsquare_indices = [(row, col) for row in range(3) for col in range(3)]
+
+    # Build full sudoku array
+    for r_scalar in range(m):
+        for c_scalar in range(m):
+            for digit in digits:
+                # Shifts for moving subsquare inside sudoku matrix
+                row_shift = r_scalar * m
+                col_shift = c_scalar * m
+
+                # Build the labels for a subsquare
+                subsquare = [get_label(row + row_shift, col + col_shift, digit)
+                             for row, col in subsquare_indices]
+                subsquare_bqm = combinations(subsquare, 1)
+                bqm.update(subsquare_bqm)
+
+    # Constraint: Fix known values
+    for row, line in enumerate(matrix):
+        for col, value in enumerate(line):
+            if value > 0:
+                bqm.fix_variable(get_label(row, col, value), 1)
```

Comment:
Could you add some explanation here why you fix the value to 1?

Reply:
That's a good question. I hadn't realized that I made a jump until you pointed it out. Thanks